dvc/external_repo.py

```diff
@@ -102,19 +125,29 @@ def __init__(
         uninitialized=False,
         **kwargs,
     ):
-        super().__init__(
-            root_dir, scm=scm, rev=rev, uninitialized=uninitialized
-        )
-
         self.url = url
         self.for_write = for_write
-        self.cache_dir = cache_dir or self._get_cache_dir()
-        self.cache_types = cache_types
-
-        self._setup_cache(self)
-        self._fix_upstream(self)
         self.tree_confs = kwargs
 
+        self._cache_config = {
+            "cache": {
+                "dir": cache_dir or self._get_cache_dir(),
+                "type": cache_types,
+            }
+        }
+
+        config = self._cache_config.copy()
+        if os.path.isdir(url):
+            config.update(_get_remote_config(url))
+
+        super().__init__(
+            root_dir,
+            scm=scm,
+            rev=rev,
+            uninitialized=uninitialized,
+            config=config,
+        )
```

Comment:
Looks like this could be unified with `make_repo` factory. But that's for the followup pr.